From 3692fd3506e6c5f9fa26d68b3df641ae404b6381 Mon Sep 17 00:00:00 2001
From: Sai Vemuri <svemuri@redhat.com>
Date: Thu, 11 Sep 2014 22:42:48 -0400
Subject: [infiniband] cxgb4: Mind the sq_sig_all/sq_sig_type QP attributes

Message-id: <1410475447-94392-30-git-send-email-svemuri@redhat.com>
Patchwork-id: 93797
O-Subject: [RHEL7.1 PATCH BZ 1124947 029/108] RDMA/cxgb4: Mind the sq_sig_all/sq_sig_type QP attributes
Bugzilla: 1124947
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Signed-off-by: Steve Wise <swise@opengridcomputing.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit ba32de9d8d8173a1d6dd1ed608c519d5d0a623bb)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/cxgb4/iw_cxgb4.h b/drivers/infiniband/hw/cxgb4/iw_cxgb4.h
index eb18f9b..bf28ef8 100644
--- a/drivers/infiniband/hw/cxgb4/iw_cxgb4.h
+++ b/drivers/infiniband/hw/cxgb4/iw_cxgb4.h
@@ -446,6 +446,7 @@ struct c4iw_qp {
  atomic_t refcnt;
  wait_queue_head_t wait;
  struct timer_list timer;
+ int sq_sig_all;
 };
 
 static inline struct c4iw_qp *to_c4iw_qp(struct ib_qp *ibqp)
diff --git a/drivers/infiniband/hw/cxgb4/qp.c b/drivers/infiniband/hw/cxgb4/qp.c
index ff1b671..cb76eb5 100644
--- a/drivers/infiniband/hw/cxgb4/qp.c
+++ b/drivers/infiniband/hw/cxgb4/qp.c
@@ -715,7 +715,7 @@ int c4iw_post_send(struct ib_qp *ibqp, struct ib_send_wr *wr,
   fw_flags = 0;
   if (wr->send_flags & IB_SEND_SOLICITED)
    fw_flags |= FW_RI_SOLICITED_EVENT_FLAG;
-  if (wr->send_flags & IB_SEND_SIGNALED)
+  if (wr->send_flags & IB_SEND_SIGNALED || qhp->sq_sig_all)
    fw_flags |= FW_RI_COMPLETION_FLAG;
   swsqe = &qhp->wq.sq.sw_sq[qhp->wq.sq.pidx];
   switch (wr->opcode) {
@@ -776,7 +776,8 @@ int c4iw_post_send(struct ib_qp *ibqp, struct ib_send_wr *wr,
   }
   swsqe->idx = qhp->wq.sq.pidx;
   swsqe->complete = 0;
-  swsqe->signaled = (wr->send_flags & IB_SEND_SIGNALED);
+  swsqe->signaled = (wr->send_flags & IB_SEND_SIGNALED) ||
+      qhp->sq_sig_all;
   swsqe->flushed = 0;
   swsqe->wr_id = wr->wr_id;
 
@@ -1601,6 +1602,7 @@ struct ib_qp *c4iw_create_qp(struct ib_pd *pd, struct ib_qp_init_attr *attrs,
  qhp->attr.enable_bind = 1;
  qhp->attr.max_ord = 1;
  qhp->attr.max_ird = 1;
+ qhp->sq_sig_all = attrs->sq_sig_type == IB_SIGNAL_ALL_WR;
  spin_lock_init(&qhp->lock);
  mutex_init(&qhp->mutex);
  init_waitqueue_head(&qhp->wait);
-- 
1.7.1